박스오피스II [STEP 1] Yetti, Maxhyunm

maxhyunm opened the pull request:
안녕하세요 델마!(@delmaSong)
Step4 수정과 두 번째 STEP1을 함께 작업해서 올려드립니다!
이번에도 리뷰 잘 부탁드립니다🙇‍♀️

## 고민했던 점
### 1. `setSelected 설정을 위한 DateComponents`
`CalendarView`에서 현재 날짜를 선택 상태로 만들기 위해 `setSelected` 메서드를 활용하려고 보니, 해당 메서드에는 `Date` 타입이 아닌 `DateComponents` 타입을 파라미터로 전달받고 있었습니다. 이에 따라 현재 `Date` 타입으로 갖고 있는 값을 어떻게 변환시켜줄지에 대한 고민이 있었습니다.
처음에는 기존에 제작한 포매터를 활용해 `Date`를 `String`으로 변환한 뒤, `"-"`를 기준으로 년/월/일로 나누어 `DateComponents(year:month:day:)` 형식으로 전달을 하려 했습니다. 하지만 이렇게 하면 불필요한 코드가 너무 늘어나는 것 같아서, 대신 `Calendar`와 `Timezone`을 활용해 아래와 같이 구현하였습니다.

```swift
let components = Calendar.current.dateComponents(in: .current, from: date)
dateSelection.setSelected(components, animated: false)
```
#### Date와 DateComponents의 차이
- `Date`는 달력이나 timezone의 영향을 받지 않는 독립적인 시간을 나타냅니다. 현재 시점의 날짜나 시간을 구할 때 사용할 수 있으며, 날짜/시간의 연산에도 활용 가능합니다.
- `DateComponents`는 날짜의 구성 요소를 캡슐화한 타입입니다. 특정 달력과 timezone에서 날짜/시간을 구성하는 요소(년/월/일 등)를 제공하여, 특정 `Date`타입을 만들어낼 수 있게 합니다. 또한 1시간 30분과 같은 기간을 구성하는 데도 사용될 수 있습니다.

Commit:
refactor: 스택뷰의 text설정메서드 기능 분리 및 Namespace 생성

maxhyunm committed:
refactor: MovieDetailView 로딩화면 수정, NetworkConfiguration 및 NetworkingManager 타입 수정

Commit:
feat: CalendarViewController추가 및 Calendar 기본 구현

maxhyunm committed:
feat: BoxOfficeViewController와 CalendarViewController 데이터 전달 구현

delmaSong commented on BoxOffice/Model/NetworkingManager.swift:
```diff
@@ -14,7 +14,23 @@ struct NetworkingManager {
+        guard let url = urlComponents?.url else {
+            return
+        }
+        
```
여기서 리턴될때도 completion으로 실패했다는 걸 넘겨야 해당 메서드를 호출하는 곳에서도 성공/실패 여부를 인지할 수 있지 않을까요?

iOS-Yetti replied:
말씀을 듣고 보니 여기서만 그냥 리턴을 하면 안 될 것 같네요! 실패 여부 넘길 수 있도록 수정하였습니다.

delmaSong commented on BoxOffice/Controller/CalendarViewController.swift:
```diff
@@ -0,0 +1,72 @@
+import UIKit
+
+final class CalendarViewController: UIViewController {
+    let date: Date
```
date 프로퍼티를 가지고 있지 않게끔도 만들 수 있지 않을까요?
그리고 외부로 공개되어야 하는 변수인지도 모르겠네요

iOS-Yetti replied:
init에서 setUpCalendarSelection 메서드를 호출하여 해결하였습니다!

delmaSong commented on BoxOffice/Controller/MovieDetailViewController.swift:
```diff
@@ -247,22 +214,34 @@ extension MovieDetailViewController: URLSessionDelegate {
+        static let genresTitle = "장르"
+        static let actorsTitle = "배우"
     }
 }
```
enum을 이용하시면서 case가 아니라 static let 변수로 할당하신 이유가 있나요?
정적 변수로 할당을 위한거라면 구조체나 다른 타입이 아니라 왜 열거형을 사용하셨는지 궁금합니다.

iOS-Yetti replied:
해당 Namespace는 각 String 값만 활용하면 되기 때문에 굳이 case 처리를 할 필요가 없다고 생각했습니다. 또 static으로 지정하면 rawValue로 접근하지 않아도 되는 점이 좋다고 생각하여 그렇게 작업하였습니다. 그런데 델마의 코멘트를 보니, 그럼 굳이 enum 타입으로 할 필요가 없었을 것 같습니다. 이 부분은 struct로 변경해 두었습니다.

delmaSong commented on BoxOffice/Controller/BoxOfficeViewController.swift:
```diff
@@ -170,3 +170,17 @@ extension BoxOfficeViewController {
+    func setUpDate(_ date: Date)
+}
+
+extension BoxOfficeViewController: BoxOfficeDelegate {
+    func setUpDate(_ date: Date) {
+        self.date = date
+
+        isLoading = true
+        self.title = dateString
+        passFetchedData()
```
`isLoading`에 대한 컨트롤은 결국 data가 fetch되기 시작하고, 완료되었을 때 변경되는거라
`DispatchQueue.main.async` 블럭 내에서 `false`로 변경하는 것 처럼 `passFetchedData()`내에서 관리하는 편이 일관성있는 흐름으로 보여지는데 어떻게 생각하세요?

iOS-Yetti replied:
isLoading 처리 전반을 passFetchedData 내부로 이동하였습니다!

delmaSong commented on BoxOffice/Controller/BoxOfficeViewController.swift:
```diff
@@ -102,9 +109,9 @@ extension BoxOfficeViewController {
     }
     
     private func setUpDataSnapshot(_ data: [BoxOfficeEntity.BoxOfficeResult.DailyBoxOffice]) {
-        var snapshot = NSDiffableDataSourceSnapshot<NetworkNamespace, BoxOfficeEntity.BoxOfficeResult.DailyBoxOffice>()
+        var snapshot = NSDiffableDataSourceSnapshot<NetworkConfiguration, BoxOfficeEntity.BoxOfficeResult.DailyBoxOffice>()
         
-        snapshot.appendSections([.boxOffice])
+        snapshot.appendSections([.boxOffice(dateString.replacingOccurrences(of: "-", with: ""))])
```
`dateString.replacingOccurrences(of: "-", with: "")` 이 부분이 두군데 이상 사용되고있는데, 뷰컨의 전역변수로 선언된 `dateString` 변수와 연관지어서 좀 더 간단하게 쓸 수도 있을 것 같네요

maxhyunm replied:
해당 부분 format 별로 날짜 String을 반환하는 메서드를 생성하여 처리하였습니다!

delmaSong commented on BoxOffice/Resource/NetworkConfiguration.swift:
```diff
@@ -0,0 +1,62 @@
+    
+    var query: [(name: String, value: String)] {
+        switch self {
+        case .boxOffice:
+            return []
+        case .movieDetail:
+            return []
```
default: return [] 로 간단하게 만들수도 있을 것 같네용

maxhyunm replied:
해당 부분 적용하였습니다!

delmaSong commented on BoxOffice/Resource/SceneDelegate.swift:
```diff
@@ -7,8 +7,8 @@
 
-@available(iOS 14.0, *)
-class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+@available(iOS 16.0, *)
```
여기 버전 지정해주는 부분이 왜 필요한건가요?

maxhyunm replied:
CalendarView가 iOS 16부터 적용되기 때문에 CalendarView를 사용하기 위해 작성해두었었습니다! 하지만 미니멈 타겟버전을 16.0으로 맞추면서 필요없는 부분이 되었네요 😅 해당 부분 삭제하였습니다!

delmaSong commented on BoxOffice/Resource/NetworkConfiguration.swift:
```diff
@@ -0,0 +1,62 @@
+    
+    var url: String {
+        switch self {
+        case .boxOffice(let targetDate):
+            return String(format: "http://kobis.or.kr/kobisopenapi/webservice/rest/boxoffice/searchDailyBoxOfficeList.json?key=%@&targetDt=%@", NetworkConfiguration.apiKey, targetDate)
```
apiKey가 없거나 잘못된 상황에 대한 검증은 어떻게 하나요?
Bundle에 해당 키가 없으면 ""가 들어갈텐데, 해당 부분에 대한 필터링은 어떻게 이뤄지나요?

iOS-Yetti replied:
지금은 Key가 없을 경우 잘못된 URL 오류, 잘못되었을 경우에는 response상에서 전달받은 오류를 보여주고 있습니다. 해당 부분 API Key 검증 차원에서 예외처리를 추가하려고 해보았는데, 그러려면 관련 연산 프로퍼티들이 모두 throws를 추가한 메서드 형식으로 바뀌어야 해서 고민이 됩니다.(예: url 프로퍼티 -> getUrl 메서드, header 프로퍼티 -> getHeader 메서드) 이 방향으로 수정하는 것이 적절한 방향인지 델마의 의견을 여쭤보고 싶습니다...!

delmaSong replied:
음.. 아주 간단하게는 이 url의 타입을 `String?`으로 변경해서 사용하는 쪽에서 제대로된 url인지 검증하는 부분에서 걸러지도록 할 수 있을 것 같네요..!

delmaSong reviewed (commented):
수고 많으셨습니다!
전반적으로 잘 진행해주셨으나, 커밋 단위가 너무 크고 커밋 메시지가 각 작업 단위를 잘 나타내지는 못하는 것 같습니다.
요 부분 조금만 더 신경써주시면 좋을 것 같아요 😃
아래 코멘트 확인해주세요~!

Commit:
refactor: 리뷰 반영하여 코드 수정

- CalendarViewController에서 date프로퍼티 삭제
- url 호출 실패시 completion으로 에러처리 추가
- Namespace enum에서 struct로 변경
- NetworkConfiguration에서 중복되는 리턴 default처리
- SceneDelegate에서 버전정보 삭제

maxhyunm committed:
refactor: 리뷰 반영하여 코드 수정(2)

- BoxOfficeViewController에서 isLoading 호출 위치 변경
- MovieDetailViewController에서 isDataLoading, isImageLoading 호출 위치 변경
- dateString을 메서드로 변환하여 포맷별 날짜 String값을 받을 수 있도록 수정

delmaSong commented:
@maxhyunm @iOS-Yetti 수고 많으셨어요! 다음 스텝에서 만나요!!
